src/host/readDataCooked.cpp

```diff
@@ -797,453 +725,452 @@ void COOKED_READ_DATA::_transitionState(State state) noexcept
+
+    std::wstring output;
+
+    // Disable the cursor when opening a popup, reenable it when closing them.
+    if (const auto popupOpened = !_popups.empty(); _popupOpened != popupOpened)
     {
-        return 0;
+        wchar_t buf[] = L"\x1b[?25l";
+        buf[5] = popupOpened ? 'l' : 'h';
+        output.append(&buf[0], 6);
+        _popupOpened = popupOpened;
     }
 
-    auto& textBuffer = _screenInfo.GetTextBuffer();
-    const auto width = textBuffer.GetSize().Width();
-    auto columnLimit = width - _getColumnAtRelativeCursorPosition(cursorOffset);
+    // Scroll the contents of the pager if needed, so we only need to write what actually changed.
+    if (const auto delta = pagerContentTop - _pagerContentTop; delta != 0)
+    {
+        const auto deltaAbs = abs(delta);
+        til::CoordType beg = 0;
+        til::CoordType end = pagerHeight;
+
+        // If the top changed by more than the viewport height, scrolling doesn't make sense.
+        if (deltaAbs < size.height)
+        {
+            beg = delta >= 0 ? pagerHeight - deltaAbs : 0;
+            end = delta >= 0 ? pagerHeight : deltaAbs;
+            const auto cmd = delta >= 0 ? L'S' : L'T';
+            fmt::format_to(std::back_inserter(output), FMT_COMPILE(L"\x1b[{}{}"), deltaAbs, cmd);
+        }
 
-    size_t offset = 0;
-    ptrdiff_t distance = 0;
+        // Mark each row that has been uncovered by the scroll as dirty.
+        for (auto i = beg; i < end; i++)
+        {
+            auto& line = lines.at(i + pagerContentTop);
+            line.dirtyBegOffset = 0;
+            line.dirtyBegColumn = 0;
+        }
+    }
 
-    while (offset < text.size())
+    bool anyDirty = false;
+    for (til::CoordType i = 0; i < pagerHeight; i++)
     {
-        til::CoordType columns = 0;
-        offset += textBuffer.FitTextIntoColumns(text.substr(offset), columnLimit, columns);
-        distance += columns;
-        columnLimit = width;
+        const auto& line = lines.at(i + pagerContentTop);
+        anyDirty = line.dirtyBegOffset < line.text.size();
+        if (anyDirty)
+        {
+            break;
+        }
     }
 
-    return distance;
-}
+    til::point writeCursorPosition{ -1, -1 };
 
-// A helper to write text and calculate the number of cells we've written.
-// _unwindCursorPosition then allows us to go that many cells back. Tracking cells instead of explicit
-// buffer positions allows us to pay no further mind to whether the buffer scrolled up or not.
-ptrdiff_t COOKED_READ_DATA::_writeCharsUnprocessed(const std::wstring_view& text) const
-{
-    if (text.empty())
+    if (anyDirty)
     {
-        return 0;
-    }
+#if COOKED_READ_DEBUG
+        static size_t debugColorIndex = 0;
+        const auto color = til::colorbrewer::dark2[++debugColorIndex % std::size(til::colorbrewer::dark2)];
+        fmt::format_to(std::back_inserter(output), FMT_COMPILE(L"\x1b[48;2;{};{};{}m"), GetRValue(color), GetGValue(color), GetBValue(color));

```

Comment:
consider ITU T.418

```suggestion
        fmt::format_to(std::back_inserter(output), FMT_COMPILE(L"\x1b[48:2::{}:{}:{}m"), GetRValue(color), GetGValue(color), GetBValue(color));
```

Reply:
Hmm... 🤔 https://github.com/microsoft/terminal/issues/17426#issuecomment-2166215583
> Although when it comes to passing the sequence through to conpty, I think we decided that semicolons are best for 38 and 48, since they're more widely supported, while colons were best for 58, since all terminals that support underline colors also support colons, but not all of them support semicolon (we obviously don't).

Should we?

Reply:
No you’re right